src/algebra/category/Group/epi_mono.lean

```diff
@@ -0,0 +1,334 @@
+import algebra.category.Group.basic
+
+/-!
+# Monomorphisms and epimorphism in `Group`
+In this file, we prove monomorphisms in category of group are injective homomorphisms and
+epimorphisms are surjective homomorphisms.
+-/
+
+
+universes u
+
+namespace monoid_hom
+
+open quotient_group
+
+variables {A B : Type u}
+
+section
+
+variables [group A] [group B]
+
+@[to_additive add_monoid_hom.ker_eq_bot_of_cancel]
+lemma ker_eq_bot_of_cancel {f : A →* B}
+  (h : ∀ (u v : f.ker →* A), f.comp u = f.comp v → u = v) :
+  f.ker = (⊥ : subgroup A) :=
+begin
+  specialize h 1 f.ker.subtype begin
+    ext1,
+    simp only [comp_one, one_apply, coe_comp, subgroup.coe_subtype, function.comp_app],
+    rw [←subtype.val_eq_coe, f.mem_ker.mp x.2],
+  end,
+  rw [←subgroup.subtype_range f.ker, ←h, range_one],
+end
+
+end
+
+section
+
+variables [comm_group A] [comm_group B]
+
+@[to_additive add_monoid_hom.range_eq_top_of_cancel]
+lemma range_eq_top_of_cancel {f : A →* B}
+  (h : ∀ (u v : B →* B ⧸ f.range), u.comp f = v.comp f → u = v) :
+  f.range = ⊤ :=
+begin
+  specialize h 1 (quotient_group.mk' _) begin
+    ext1,
+    simp only [one_apply, coe_comp, coe_mk', function.comp_app],
+    rw [show (1 : B ⧸ f.range) = (1 : B), from quotient_group.coe_one _, quotient_group.eq, inv_one,
+      one_mul],
+    exact ⟨x, rfl⟩,
+  end,
+  replace h : (quotient_group.mk' _).ker = (1 : B →* B ⧸ f.range).ker := by rw h,
+  rwa [ker_one, quotient_group.ker_mk] at h,
+end
+
+end
+
+end monoid_hom
+
+section
+
+open category_theory
+
+namespace Group
+
+variables {A B : Group.{u}} (f : A ⟶ B)
+
+@[to_additive AddGroup.ker_eq_bot_of_mono]
+lemma ker_eq_bot_of_mono [mono f] : f.ker = ⊥ :=
+monoid_hom.ker_eq_bot_of_cancel $ λ u v,
+  (@cancel_mono _ _ _ _ _ f _
+    (show Group.of f.ker ⟶ A, from u) _).1
+
+@[to_additive AddGroup.mono_iff_ker_eq_bot]
+lemma mono_iff_ker_eq_bot :
+  mono f ↔ f.ker = ⊥ :=
+⟨λ h, @@ker_eq_bot_of_mono f h,
+λ h, concrete_category.mono_of_injective _ $ (monoid_hom.ker_eq_bot_iff f).1 h⟩
+
+@[to_additive AddGroup.mono_iff_injective]
+lemma mono_iff_injective :
+  mono f ↔ function.injective f :=
+iff.trans (mono_iff_ker_eq_bot f) $ monoid_hom.ker_eq_bot_iff f
+
+namespace surjective_of_epi_auxs
+
+local notation `X` := set.range (function.swap left_coset f.range.carrier)
+
+/--
+Define `X'` to be the set of all right cosets with an extra point at "infinity".
+-/
+@[nolint has_inhabited_instance]
+inductive X_with_infinity
+| from_coset : set.range (function.swap left_coset f.range.carrier) → X_with_infinity
+| infinity : X_with_infinity
+
+
+open X_with_infinity equiv.perm
+open_locale coset
+
+local notation `X'` := X_with_infinity f
+local notation `∞` := X_with_infinity.infinity
+local notation `SX'` := equiv.perm X'
+
+instance : has_smul B X' :=
+{ smul := λ b x, match x with
+  | from_coset y := from_coset ⟨b *l y, begin
+    rw [←subtype.val_eq_coe, ←y.2.some_spec, left_coset_assoc],
+    use b * y.2.some,
+  end⟩
```

Comment:
Here, and in the rest of the file, please place `begin`s in their own line and adjust the indentation like this:
```suggestion
  | from_coset y := from_coset ⟨b *l y,
    begin
      rw [←subtype.val_eq_coe, ←y.2.some_spec, left_coset_assoc],
      use b * y.2.some,
    end⟩
```